chore(build-system) Full (cross) build inside a main container

1138-4EB opened the pull request:
This PR is major rewrite of the build procedure, which can have impact in how developers iteratively write/build/test. However, it provides a host-agnostic environment, so that bugs/errors/features can be directly reproduced.

# Features

List of features already integrated to `develop` which work both, locally (usual workflow) and inside `portainer/factory` (portable workflow):

- [ ] Portable build system (no connection required)
- [ ] Multi-platform support (those with docker static binaries available)
- Run `gruntfile.js` (#1335)
  - `lint`
  - `frontend`
    - [ ] `frontend:release`
    - [ ] `frontend:dev`
  - [ ] `shell:downloadDockerBinary:<platform>:<arch>`
  - [ ] `shell:extractDockerBinary:<platform>:<arch>`
  - [ ] `build:<platform>:<arch>`
  - [ ] `release:<platform>:<arch>`
  - [ ] `run-dev`
- Run `build.sh`
  - [ ] `v0.0.0`
    - [ ] ca-certificates
    - [ ] Frontend
    - [ ] Backend
    - [ ] Docker static binary
    - [ ] Multiple artifacts (tarballs/images)
- [ ] Integrated multi-arch image manifest management.
  - [portainer/docker-manifest](https://github.com/portainer/docker-manifest)
  - [estesp/manifest-tool](https://github.com/estesp/manifest-tool)
  - [docker/cli#138](https://github.com/docker/cli/pull/138)

# Tasks

- Reduce dependencies
  - [ ] portainer/golang-builder (#1333)
  - [ ] portainer/base
- [ ] Check if platform is supported for development. (#1330)
- Allow to modify the paths at runtime when executing `grunt`:
  - [ ] Add `workdir` config variable to the `gruntfile`. Optionally set it with an environment variable, `WORKDIR`. If `WORKDIR` is empty, default to `./`. (#1330)
  - [ ] Allow empty `distdir`. (#1330)
- [ ] Split docker binary download and extract steps to two grunt tasks. Drop `download_docker_binary.sh`. (#1331)
- [ ] Split `grunt run` to `grunt rm` and `grunt run`. Add check to `grunt rm`. (#1331)
- [ ] Watch `api` and `app` separately. (#1330)
  - [ ] Build and update frontend with changes in `app`.
  - [ ] Build and update backend with changes in `api`.
- `portainer/builder` image (#1333)
  - [ ] Replace `portainer/golang-builder` with a script combined with `golang:alpine`.
  - [ ] Enhance the script to allow saving dependencies with `godep`. Allow to only save them, without actually building the pkg.
  - [ ] Add `git` to a `golang` image to allow getting dependencies.
  - [ ] Add `upx` to a `golang` image to allow compression of produced binaries.
- `portainer-factory` image
  - [ ] Get golang dependencies with `portainer/builder` and save them to `/portainer-deps/vendor` and `/portainer-deps/Godeps`.
  - [ ] Install `nodejs`, `nodejs-npm`, and `dos2unix`.
  - [ ] Install npm/bower dependencies.
  - [ ] Remove no longer needed dependencies: `npm`, `git`.
  - [ ] Install/get `docker`.
  - [ ] Clear temporal files.
  - [ ] Get docker static binaries with `grunt` and save them to `/docker_binaries`. (#1334)


# Context

Yesterday, I wanted to build the branch in #900, in order to reply to the comment. Since, I was not using the same machine from where it was opened, I had no dependencies installed (nodejs, grunt...). On top of that, I was in a Windows 10 x64 box, with docker 17.03.1-ce and MINGW64_NT-10.0 2.7.0(0.306/5/3) x86_64. At now, npm cannot be installed with pacman (mingw64): [Alexpub/MINGW-packages/#2249](https://github.com/Alexpux/MINGW-packages/issues/2249).

Therefore, I decided to adapt the procedure so that all the dependencies are installed in a docker image, which I named `portainer/factory`. At the same time, I felt that `build.sh` and `gruntfile.js` where too verbose.

# A data volume

The main 'problem' with the actual build procedure is that `build*Binary` grunt tasks (of type `shell`) run separate docker containers with locally mounted volumes:

``` javascript
buildWindowsAmd64Binary: {
        command: [
          'docker run --rm -v $(pwd)/api:/src -e BUILD_GOOS="windows" -e BUILD_GOARCH="amd64" portainer/golang-builder:cross-platform /src/cmd/portainer',
...
```

Then, if grunt is run inside a container with the host's `/var/run/docker.sock` mounted as a volume, `$(pwd)` is read in the container but the path which is actually mounted corresponds to the host: [moby/moby/#28124](https://github.com/moby/moby/issues/28124). The result is that `/src` is empty in `portainer/golang-builder` instances, and the build fails.

The workaround, as explained in the issue above, is to use a named voume. However, docker does not support directly mounting a path in the host to a named volume. The suggested approach is to mount it in a container and execute `mv` or `cp` inside: https://docs.docker.com/engine/tutorials/dockervolumes/#backup-restore-or-migrate-data-volumes

Then, the first modification to the gruntfile is to replace `$(pwd)` with a name. I took `portainer-src-volume`.

# Documentation

A draft to adapt the documentation to reflect changes introduced by this PR is available at https://github.com/portainer/portainer-docs/pull/29. In the following sections it is intensively referred to.

# `portainer/factory` and `portainer/builder`

The portable build system is composed of two images. These can be made available at portainers docker hub, or any user can generate them from a work copy of this git repo running `./build/factory_create.sh` (calling the script from any other path should also work). See [contribute.rst#factory-scripts](https://github.com/1138-4EB/portainer-docs/blob/b22b7421647dcc16b573deef2c0b82623c2085f8/docs/source/contribute.rst#factory-scripts):

- Build image `portainer/factory`: [factory_create.sh#L7-L15](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/factory_create.sh#L7-L15)
  - It is just `golang:alpine` with `git` and `upx`.
- Download golang dependencies to a local path: [factory_create.sh#L20-L28](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/factory_create.sh#L20-L28)
  - Note that `portainer-factory` and [`builder.sh`](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/builder.sh) are used. `builder.sh` is just an adaptation of the one at [portainer/golang-builder](https://github.com/portainer/golang-builder) that allows us to drop `portainer/golang-builder`.
- Build image `portainer/factory`: [factory_create.sh#L31-L48](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/factory_create.sh#L31-L48)
  - [This](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/Dockerfile) Dockerfile is used.
    - Golang dependencies are copied. Note that this requires `portainer/factory` to be rebuilt each time any of these change.
    - `package.json` and bower.json` are copied. Note that this requires `portainer/factory` to be rebuilt each time any of these change.
    - `nodejs`, `grunt`, `bower`, `docker`, `dos2unix`, `git` and `rsync` (I think that `git` could actually be dropped) are installed.
    - The entrypoint is set to [`bootstrap.sh`](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/bootstrap.sh). See [/contribute.rst#factory-image](https://github.com/1138-4EB/portainer-docs/blob/b22b7421647dcc16b573deef2c0b82623c2085f8/docs/source/contribute.rst#factory-image). To understand why `bootstrap.sh` is required, note that each time `portainer/factory` is executed during development, the local path of the git repo will be bind to `/work`. But all the processes will be run from `WORKDIR`, which defaults to [/portable-build](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/Dockerfile#L9). Last, as explained above, a named data volume is required to shared data between the `portainer/factory` container (named `portainer-factory` during execution) and `golang/builder` container (to build the backend(s)). This data volume, named `portainer-src-volume` is bind to `/src-volume` in `portainer-factory` and to `/src` in builders.
- Download docker static binaries for all the supported platforms: [factory_create.sh#L53-L67](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/factory_create.sh#L53-L67)
- Add docker binaries to `portainer-factory` at `/docker-binaries`: [factory_create.sh#L70-L79](https://github.com/1138-4EB/portainer/blob/aea0ff6257be1f014f81083846fc8e427d6760d9/build/factory_create.sh#L70-L79)

Note that anyone who either pulls or creates `portainer/builder` and `portainer/factory` can develop portainer, and even generate artifacts (tarballs and/or images) offline. Those, along with a copy of the git repo are the "portable build system". Images are platform agnostic, so that the same image can be used in Linux, Windows and Mac. This is interesting because it provides an standard development setup for all the contributors.

1138-4EB commented:
I forgot to mention that in order to use `build_env.sh` in MINGW, the line where portainer-builder is run needs to be replaced with:

```
winpty docker run -itv //portainer-src-volume:/src-volume -v //var/run/docker.sock:/var/run/docker.sock --name portainer-main-builder portainer/portainer-builder "$@"
```

This line is commented in the file.

deviantony commented:
Looks really interesting !! I'll need to take some time to review/test this tho, will try as soon as I can but I've got other priorities at the moment.

We have an issure related to gruntfile refactoring: https://github.com/portainer/portainer/issues/40, I'll link so we can track this.

I've seen the `recess` issue before, it's probably due to the fact that the `bower_components/` directory is not present... are you sure that the bower dependencies are correctly installed?

Quick answers to your questions/issues:

* Issue 1: yes, the artifacts generated by the build (go binary and javascript app) should be available from outside the container (login and pushing is related to 'deployment' and not building, so it should be considered outside from this process/PR) 
* Issue 2: why would the container requires access to the host Docker socket?
* Issue3: see my comment above for the recess problem, I think that you're missing a `bower install` step after `npm install`

My two cents without entirely reviewing the PR: a portable build system should provide all the required libraries/dependencies in the image and not download them each time.

Thanks again for the work !

1138-4EB commented:
> We have an issure related to gruntfile refactoring: #40, I'll link so we can track this.

I'll open a new PR to work on the refactorization of the gruntfile. Then, I'll make this PR be based on that. I think this will help you with review, when you have time.

> Issue 1: yes, the artifacts generated by the build (go binary and javascript app) should be available from outside the container (login and pushing is related to 'deployment' and not building, so it should be considered outside from this process/PR)

I think that we are mixing 'artifacts' here. Those you refer to are saved to `/tmp/portainer-builds` (the path you use now, I didn't change it) inside the portainer-builder, and then copied to the host. This is the content of the destination dir in the host after the execution of `./build_env`:

``` bash
portainer-work/builds/
├── darwin-amd64
│   └── portainer
│       ├── css
│       │   └── app.3b18a74f.css
│       ├── fonts
│       │   ├── fontawesome-webfont.svg
│       │   ├── fontawesome-webfont.ttf
│       │   ├── fontawesome-webfont.woff
│       │   ├── fontawesome-webfont.woff2
│       │   ├── glyphicons-halflings-regular.svg
│       │   ├── glyphicons-halflings-regular.ttf
│       │   ├── glyphicons-halflings-regular.woff
│       │   ├── glyphicons-halflings-regular.woff2
│       │   ├── montserrat-regular-webfont.svg
│       │   ├── montserrat-regular-webfont.ttf
│       │   └── montserrat-regular-webfont.woff
│       ├── ico
│       │   ├── apple-touch-icon-precomposed.png
│       │   └── favicon.ico
│       ├── images
│       │   ├── logo_alt.png
│       │   └── logo.png
│       ├── index.html
│       ├── js
│       │   └── app.52f70026.js
│       └── portainer-darwin-amd64
├── linux-386
    └── ...
├── linux-amd64
    └── ...
├── linux-amd64
    └── ...
├── linux-arm
    └── ...
├── linux-arm64
    └── ...
├── linux-ppc64le
    └── ...
└── windows-amd64
    └── ...
├── portainer-R-darwin-amd64.tar.gz
├── portainer-R-linux-386.tar.gz
├── portainer-R-linux-amd64.tar.gz
├── portainer-R-linux-arm64.tar.gz
├── portainer-R-linux-arm.tar.gz
├── portainer-R-linux-ppc64le.tar.gz
├── portainer-R-windows-amd64.tar.gz
```

Note that "R" is the `<VERSION>` I used.

The artifacts I meant are the docker images which are built during the build process. More precisely, in function `build_and_push_images()` inside `build.sh`. This function is only run for 'linux' platforms. This also something I just kept (although partially rewritten, it is functionaly equivalent):

``` bash
function build_and_push_images() {
  docker build -t "portainer/portainer:$1-${VERSION}" -f build/linux/Dockerfile .
  docker tag  "portainer/portainer:$1-${VERSION}" "portainer/portainer:$1"
  docker push "portainer/portainer:$1-${VERSION}"
  docker push "portainer/portainer:$1"
}
```
The first command is properly executed, and the docker image (with two tags) is available in the host, because the socket is mounted. However, the last two commands (push), which should upload the new images to docker hub, shows an authentication error (indeed, I do not login at all). As a result, after running `build_env.sh` 5 images and 10 tags are shown in the host:

``` bash
$ docker images
REPOSITORY           TAG             IMAGE ID      SIZE
portainer/portainer  linux-ppc64le   ee2d63245487  10.27 MB
portainer/portainer  linux-ppc64le-R ee2d63245487  10.27 MB
portainer/portainer  linux-arm64     b1f8df503da2  10.07 MB
portainer/portainer  linux-arm64-R   b1f8df503da2  10.07 MB
portainer/portainer  linux-arm       d3c04a2fcd3e  9.49 MB
portainer/portainer  linux-arm-R     d3c04a2fcd3e  9.49 MB
portainer/portainer  linux-386       da8182559586  8.946 MB
portainer/portainer  linux-386-R     da8182559586  8.946 MB
portainer/portainer  linux-amd64     df6c5166f455  10.26 MB
portainer/portainer  linux-amd64-R   df6c5166f455  10.26 MB
```

According to your last message I'd say that `docker push` commands should be dropped from `build.sh`. Then, a new script (named `deploy.sh`?) could be added to i) push images and ii) push tarballs. However, this would break the current behaviour of `build.sh`.

> Issue 2: why would the container requires access to the host Docker socket?

Ignoring the modifications I did, i.e., in your current dev branch, `build<Platform><Arch>Binary` grunt tasks execute a command to run a docker container:

```` javascript
buildWindowsAmd64Binary: {
        command: [
          'docker run --rm -v $(pwd)/api:/src -e BUILD_GOOS="windows" -e BUILD_GOARCH="amd64" portainer/golang-builder:cross-platform /src/cmd/portainer',
...
````

Therefore, if grunt itself is run inside a container (`portainer/portainer-builder`), all those `portainer/golang-builder:cross-platform` instances are created from a container. As far as I am aware, the most straightforward approach to allow so is to mount the socket: https://jpetazzo.github.io/2015/09/03/do-not-use-docker-in-docker-for-ci/

On top of that, these are two major advantages over using docker-in-docker:

- Built and tagged docker images are directly available in the host. This is desirable to split build from deploy, as commented above.
- Docker image `portainer/golang-builder:cross-platform` will be kept in the host. With docker-in-docker, it would be downloaded every time the `portainer/portainer-builder` is created.

> My two cents without entirely reviewing the PR: a portable build system should provide all the required libraries/dependencies in the image and not download them each time.

The current proposal is host-agnostic, not portable. That is, the purpose is to run https://portainer.readthedocs.io/en/stable/contribute.html in a container, not to be run offline. This is because the sources are cloned, and docker images and pulled, so connection is required anyway. Nevertheless, since local sources can be used, and docker images can be exported/imported, it is sensible to make it portable:

As far as npm and bower are concerned, these are installed and run when the `portainer-builder` is built. Therefore, the JavaScript part is portable.

However, the golang part is not portable, because dependencies are downloaded each time a container based on `portainer/golang-builder:cross-platform` is built. I can have a look at it, but note that I didn't touch `portainer/golang-builder:cross-platform` at all. I am afraid that copying golang dependencies to `/src` is not  enough, so the image might require modifications.

The approach I am considering is to install all the golang dependencies in `portainer-builder`, just as I did with `node_modules` and `bower_components`. There is a subtle difference tho, since the latter two are used in `portainer-builder` itself, but golang dependencies must be made available to `portainer/golang-builder:cross-platform` containers. As explained in the first message, a named docker volume is required. Therefore, if `/src` cannot be reused for both the source and the dependencies, two named volumes must be used.

Which approach do you prefer? Shall I use two named containers, or will I modify `golang-builder:cross-platform` so that it looks for both sources and dependencies in `/src`?

> Issue3: see my comment above for the recess problem, I think that you're missing a bower install step after npm install

Indeed, bower was not installed at all. I added it, following the same approach as with `node_modules`. It works seamlessly now. These are the relevant commands in the dockerfile:

``` bash
wget https://raw.githubusercontent.com/$REPO/$BRANCH/package.json
wget https://raw.githubusercontent.com/$REPO/$BRANCH/bower.json
npm install -g bower findup-sync resolve nopt grunt-known-options && npm install
bower install --allow-root
````

Still, the following warnings are shown, but I think that none is critical:

``` bash
npm WARN deprecated bower@1.8.0: ..psst! While Bower is maintained, we recommend Yarn and Webpack for *new* front-end projects! Yarn's advantage is security and reliability, and Webpack's is support for both CommonJS and AMD projects. Currently there's no migration path but we hope you'll help us figure out one.

npm WARN deprecated bower@1.8.0: ..psst! While Bower is maintained, we recommend Yarn and Webpack for *new* front-end projects! Yarn's advantage is security and reliability, and Webpack's is support for both CommonJS and AMD projects. Currently there's no migration path but we hope you'll help us figure out one.
npm WARN deprecated grunt-recess@0.3.5: Deprecated as RECESS is unmaintained
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated minimatch@0.3.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated graceful-fs@1.2.3: graceful-fs v3.0.0 and before will fail on node releases >= v7.0. Please update to graceful-fs@^4.0.0 as soon as possible. Use 'npm ls graceful-fs' to find it in the tree.
npm WARN deprecated minimatch@0.4.0: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN prefer global LiveScript@1.0.1 should be installed with -g
npm WARN prefer global istanbul@0.1.46 should be installed with -g
npm WARN prefer global jshint@2.9.4 should be installed with -g
npm WARN prefer global ycssmin@1.0.1 should be installed with -g
npm WARN prefer global recess@1.1.9 should be installed with -g
npm WARN prefer global karma@0.8.8 should be installed with -g

npm WARN lifecycle portainer@1.13.2~postinstall: cannot run in wd %s %s (wd=%s) portainer@1.13.2 bower install /portainer-build/portainer

npm WARN portainer@1.13.2 No description
npm WARN portainer@1.13.2 No license field.

bower angular-utils-pagination#~0.11.1     invalid-meta for:/tmp/db64d55248fae74019f8070a2fa031f3/bower/7e1bb4ea8a21d16f4f912addaa207805-6864-y55zhS/bower.json
bower angular-utils-pagination#~0.11.1     invalid-meta The "name" is recommended to be lowercase, can contain digits, dots, dashes

bower angular-google-analytics#~1.1.9      invalid-meta for:/tmp/db64d55248fae74019f8070a2fa031f3/bower/f41b7351bf79de70b7910978744e7bc7-6864-pFOQgi/bower.json
bower angular-google-analytics#~1.1.9      invalid-meta The "main" field cannot contain minified files

bower Chart.js#1.0.2                       invalid-meta for:/tmp/db64d55248fae74019f8070a2fa031f3/bower/671a34f677b2295c20efbb71687965ad-6864-QWAGm9/bower.json
bower Chart.js#1.0.2                       invalid-meta The "name" is recommended to be lowercase, can contain digits, dots, dashes

bower rdash-ui#1.0.*                       invalid-meta for:/tmp/db64d55248fae74019f8070a2fa031f3/bower/02e9d3bc79a7cd01135279ad336a39ff-6864-kZED16/bower.json
bower rdash-ui#1.0.*                       invalid-meta The "main" field cannot contain minified files
bower rdash-ui#1.0.*                       invalid-meta The "main" field cannot contain font, image, audio, or video files
bower rdash-ui#1.0.*                       invalid-meta The "main" field cannot contain font, image, audio, or video files
bower rdash-ui#1.0.*                       invalid-meta The "main" field cannot contain font, image, audio, or video files
bower rdash-ui#1.0.*                       invalid-meta The "main" field cannot contain font, image, audio, or video files
bower rdash-ui#1.0.*                       invalid-meta The "main" field has to contain only 1 file per filetype; found multiple .css files: ["dist/css/rdash.css","dist/css/rdash.min.css"]

bower angular                          extra-resolution Unnecessary resolution: angular#1.5.11
```

1138-4EB commented:
I moved modifications and comments related to the refactorization of the gruntfile to #939. Then I based this branch on that. However, because this PR is pointing to `develop`, all the modifications are shown here (those corresponding to #939 and #938). Once #939 is merged, these will be the modifications:

https://github.com/1138-4EB/portainer/compare/refactor40-grunt-tasks...1138-4EB:containerizedbuild

1138-4EB commented:
This last commit offers a fully portable build system. That is, golang dependencies are downloaded just once, along with `node_modules` and `bower_components`. On top of that, the whole process is based on `alpine`, instead of `debian`.

- `build.sh` is moved to `build/build.sh`
- `build/build_cross-compilation.sh` is renamed to `build/build_in_container.sh`
- The dependency on `portainer/golang-builder` is dropped. `build.sh` and `build_environment.sh` in that repo have been merged in a single file named `build/builder.sh` in this repo. This file contains the `godep` procedures as well. Then, it can be used to download depencencies only, or to build portainer. Note that it is functionally equivalent to `golang-builder`. Then, dependencies will be downloaded at build time when they are not provided through `/src`.

The main container I named `portainer-builder` in previous messages is now named `portainer/factory`. This is the build procedure:

``` bash
build_factory() {
  mkdir buildimg && cd buildimg
  mkdir godeps && mkdir tmp && cd tmp
  git clone -b feat938-portable-build http://github.com/1138-4EB/portainer ./  
    
  docker run --rm -it \
    -w /src \
    -v $(pwd)/api:/src \
    -v $(pwd)/../godeps:/godeps \
    golang:alpine sh -c "apk add -U --no-cache git; mainPackagePath=/src/cmd/portainer; DEPSONLY=true `cat ./build/builder.sh`"

  cd .. && rm -rf tmp
  wget https://raw.githubusercontent.com/1138-4EB/portainer/feat938-portable-build/build/Dockerfile
  docker build -t portainer/factory .
  cd .. && rm -rf buildimg
}
```

To make a quick test of the whole system:

``` bash
wget https://raw.githubusercontent.com/1138-4EB/portainer/feat938-portable-build/build/build_env.sh
chmod +x build_env.sh
./build_env 'factory'
./build_env 'MY.VERSION'
```
The result will be a folder named `builds` (contaning the artifacts); and five docker images available in the host (along with `portainer-factory`):

```
REPOSITORY           TAG                        IMAGE ID      SIZE
portainer/factory    latest                     22c4bfcb3924  353.6 MB
portainer/portainer  linux-ppc64le              08f38888019c  10.76 MB
portainer/portainer  linux-ppc64le-MY.VERSION   08f38888019c  10.76 MB
portainer/portainer  linux-arm64                c257ba2c9d7a  10.76 MB
portainer/portainer  linux-arm64-MY.VERSION     c257ba2c9d7a  10.76 MB
portainer/portainer  linux-arm                  a8ddb4a45369  10.76 MB
portainer/portainer  linux-arm-MY.VERSION       a8ddb4a45369  10.76 MB
portainer/portainer  linux-386                  cf1dec5076c4  10.34 MB
portainer/portainer  linux-386-MY.VERSION       cf1dec5076c4  10.34 MB
portainer/portainer  linux-amd64                65d896c9b13a  10.34 MB
portainer/portainer  linux-amd64-MY.VERSION     65d896c9b13a  10.34 MB
```

1138-4EB commented:
On the one hand, I modified some of the files so that `REPO` and `BRANCH` are provided through environment variables.
On the other hand, I added a `.travis.yml` file, so that all the binaries and images are built after each commit/PR. All of it is done in less than 9min: https://travis-ci.org/1138-4EB/portainer/builds/248611716
Moreover, artifacts can be automatically pushed to github releases and/or docker hub. However, I don't know if you want to use this feature.

deviantony commented:
I'm still looking at this, quick thoughts:

> According to your last message I'd say that docker push commands should be dropped from build.sh. Then, a new script (named deploy.sh?) could be added to i) push images and ii) push tarballs. However, this would break the current behaviour of build.sh.

Yes, I'd rather see a new `deploy.sh` script (push images would be enough for now, though pushing the tarballs on github would be nice too indeed).

> On the other hand, I added a .travis.yml file, so that all the binaries and images are built after each commit/PR. All of it is done in less than 9min: https://travis-ci.org/1138-4EB/portainer/builds/248611716
Moreover, artifacts can be automatically pushed to github releases and/or docker hub. However, I don't know if you want to use this feature.

I'd be keen to have a look at this. I've got a few questions though:
* Can it automatically creates and push an image to the hub when a PR is open?
* Can it refresh this image when any commits are made for a that PR?
* Can it delete that image once the PR is closed?
* Can it automatically creates and push an image after a commit in the develop branch ?

I'd love to use the automatic push feature for artifacts, that'd be really handy for the release process.

I reckon that I should have a look at #939 first.

1138-4EB commented:
>Yes, I'd rather see a new deploy.sh script (push images would be enough for now, though pushing the tarballs on github would be nice too indeed).

That's cool. I was not confortable with leaving those `docker tag` and `docker push` in `build.sh`. I'll move them to a new `deploy.sh`.

>EDIT: See my update on Travis below to see if Travis could be used instead of this deploy script.

Travis is just a Ubuntu (virtual) machine and it does not provide built-in features to deploy docker images. That is, the script will be executed in travis. The main point of this PR is to have a portable set of scripts so that these can be executed equally in your machine, in travis or in any other. Indeed, I used the most generic clauses, in order to make them compatible with bash, dash or the shell in alpine linux.

>I'd be keen to have a look at this. I've got a few questions though:
>    - Can it automatically creates and push an image to the hub when a PR is open?
>    - Can it refresh this image when any commits are made for a that PR?
>    - Can it delete that image once the PR is closed?
>    - Can it automatically creates and push an image after a commit in the develop branch ?

On the one hand, a docker-hub user is required (I'd recommend to create a new one just for this), and you should use travis' CLI (https://github.com/travis-ci/travis.rb#installation) to configure your credentials:

```
$ travis encrypt DOCKER_EMAIL=email@domain.com
$ travis encrypt DOCKER_USER=username
$ travis encrypt DOCKER_PASS=password
```
Then, this should be added to `.travis.yml`:

```
env:
  global:
    - secure: "UkF2CHX0lUZ...VI/LE=" # DOCKER_EMAIL
    - secure: "Z3fdBNPt5hR...VI/LE=" # DOCKER_USER
    - secure: "F4XbD6WybHC...VI/LE=" # DOCKER_PASS
```
As a result, `docker push` commands will work. For more details, see https://sebest.github.io/post/using-travis-ci-to-build-docker-images/

On the other hand, By default, travis builds are triggered with each commit to any branch, including PRs. Then, with the changes proposed here, 1, 2 and 4 are already done. I just have to write `deploy.sh` and make travis execute it on successful builds. Note that travis provides envvars to identify the branch. These can be used to set the name of the image which is pushed. See `TRAVIS_BRANCH`: https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables

Question 3 is a little tricky tho. How do we know that a PR is closed? If it is successfully merged, I suppose that we can parse `TRAVIS_COMMIT_MESSAGE` to identify it. Then, on top of pushing the new image of the destination branch, it would delete the image of the source branch. Still, there are a couple of problems:

- I am not sure about docker CLI allowing to delete images from a repository. Aparently, only `push` and `pull` are provided. The documentation tells to delete images through the GUI: https://docs.docker.com/datacenter/dtr/2.2/guides/user/manage-images/pull-and-push-images/#push-the-image 
- What should we do when a PR is closed but not merged (say it is not accepted)? I don't think that this will trigger anything in travis. Travis is added to your GitHub repo as a service, so you can not explictly set webhooks for any event you want. Well, you might be able to (see ref to GitHub API below), but it is not so straightforward.

> I'd love to use the automatic push feature for artifacts, that'd be really handy for the release process.

This is the main use case for travis, so it is a built-in feature: https://docs.travis-ci.com/user/deployment/releases/

Note these commented lines at the end of `.travis.yml`:

```
#deploy:
#  overwrite: true
#  provider: releases
#  skip_cleanup: true
#  api_key:
#    secure: 
#  file_glob: true
#  file: "*.tgz"
#  on:
#    tags: true
#    all_branches: true    
```

Ys, it is already done. You just have to use travis cli to activate the repo. It will provide you the `api_key: secure:`. As soon as you add it (and uncomment those lines), all `*.tgz` artifacts will be deployed to GitHub releases after each successful build of a *tagged* commit. Note that I set it to `all_branches`, but you can limit it to `master`.

You should take into account that a new release is created in GitHub for each deployment from travis. That is, if you 'update' a tag to point to a later commit, two releases will be shown in GitHub. The previous one will be untagged, but artifacts will not be deleted. Because of this same limitation, you can deploy untagged commits, but these will create new releases in GitHub, with 'ugly' tag names automatically created by travis. Note that no tags are created in git, but shown in the releases tab.

For completeness, travis' built-in procedure can be ignored by triggering a script the deploy, which would talk to GitHub API and manipulate releases at will. This is a workaround for the limitation, which would allow to have 'nightly' builds. I.e., once a day/week the 'develop' branch is deployed, the 'nightly' release is removed from GitHub and a new one is added. However I have never gone so far. https://github.com/conventional-changelog/conventional-github-releaser

PS: you are not required to install travis cli on your machine, you can use a docker container and just drop it when you are done.

deviantony commented:
> EDIT: docker build and docker tag steps should be part of build.sh or deploy.sh?

I reckon these steps should be part of the build.

Also, thanks a lot for your work and explanations about Travis. We're going to keep this part aside for now as we're currently using Codefresh as a CI tool. The only missing feature in Codefresh would be the automatic feature push, we might be able to do that as a script (using Github API) later on.

1138-4EB commented:
> I reckon these steps should be part of the build.

OK.

> Also, thanks a lot for your work and explanations about Travis. We're going to keep this part aside for now as we're currently using Codefresh as a CI tool. The only missing feature in Codefresh would be the automatic feature push, we might be able to do that as a script (using Github API) later on.

I had not seen the `codefresh.yml` file in the repo. I believe that it makes little sense to use Travis and Codefresh at the same time, since they provide nearly the same functionality. The first one seems to be focused on deploying artifacts, while the latter is meant to deploy docker images.

# Notes for development

On the one hand, `codefresh.yml` is, at now, composed of these steps:

- `build_backend` [in a container]
- `build_frontend` [`grunt build-webapp` in a container]
- `build_image` [local]
- `push_image` [to docker hub]

On the other hand, the proposed `.travis.yml` includes a single step, which is to execute `build/build_env.sh`. Then, `build_env.sh` should be roughly equivalent to `codefresh.yml`. However, I modified the flow to download dependencies only once. So the steps in `build_env.sh` are:

- Get golang dependencies [in a container]
- Build portainer/factory [local]
- Build portainer/builder [local]
- Run portainer/factory
  - For each target, build backend and frontend [`grunt release:<plaform>:<arch>`]
  - For linux targets, build image [local]

Comments:

It should be quite straightforward to rewrite `codefresh.yml`, so that `build_env.sh` is not used in that environment. However, there is a subtle difference: i'd say that `codefresh.yml` builds for a single platform on each run. @deviantony, can you confirm this?

In `codefresh.yml` the frontend is built once with `grunt build-webapp`. However, in `build.sh` `grunt release:...` is run once for each target (platform-arch). What is your 'real' deploy flow?
> I reckon these steps should be part of the build.

OK.

> Also, thanks a lot for your work and explanations about Travis. We're going to keep this part aside for now as we're currently using Codefresh as a CI tool. The only missing feature in Codefresh would be the automatic feature push, we might be able to do that as a script (using Github API) later on.

I had not seen the `codefresh.yml` file in the repo. I believe that it makes little sense to use Travis and Codefresh at the same time, since they provide nearly the same functionality. The first one seems to be focused on deploying artifacts, while the latter is meant to deploy docker images.

# Notes for development

On the one hand, `codefresh.yml` is, at now, composed of these steps:

- `build_backend` [in a container]
- `build_frontend` [`grunt build-webapp` in a container]
- `build_image` [local]
- `push_image` [to docker hub]

On the other hand, the proposed `.travis.yml` includes a single step, which is to execute `build/build_env.sh`. Then, `build_env.sh` should be roughly equivalent to `codefresh.yml`. However, I modified the flow to download dependencies only once. So the steps in `build_env.sh` are:

- Get golang dependencies [in a container]
- Build portainer/factory [local] (which also includes node/frontend dependencies)
- Build portainer/builder [local]
- Run portainer/factory
  - For each target, build backend and frontend [`grunt release:<plaform>:<arch>`]
  - For linux targets, build image [local]

Comments:

It should be quite straightforward to rewrite `codefresh.yml`, so that `build_env.sh` is not used in that environment. However, there is a subtle difference: i'd say that `codefresh.yml` builds for a single platform on each run. @deviantony, can you confirm this?

In `codefresh.yml` the frontend is built once with `grunt build-webapp`. However, in `build.sh` `grunt release:...` is run once for each target (platform-arch). What is your 'real' flow? Do you use 'grunt:release'?

The main point is that I might modify this PR so that the frontend is built only once (I need you to confirm that this is not platform-dependent). Then, the steps would be:

- Get golang dependencies [in a container]
- Build portainer/factory [local] (which also includes node/frontend dependencies)
- Build portainer/builder [local]
- Run portainer/factory
  - Build frontend [`grunt build-webapp`]
  - For each target, build backend and pack artifacts [`grunt release:<plaform>:<arch>`]
  - For linux targets, build image [local]

deviantony commented:
> It should be quite straightforward to rewrite codefresh.yml, so that build_env.sh is not used in that environment. However, there is a subtle difference: i'd say that codefresh.yml builds for a single platform on each run. @deviantony, can you confirm this?

Indeed, I'll take care of re-writing the `codefresh.yml` file after merge. And you're right, at the moment, Codefresh only builds a Linux AMD64 image when a new commit is pushed to the develop branch. This allows us to have the `portainer/portainer:develop` image with the latest work available.

> In codefresh.yml the frontend is built once with grunt build-webapp. However, in build.sh grunt release:... is run once for each target (platform-arch). What is your 'real' flow? Do you use 'grunt:release'?

I'm pretty sure I added the `build-webapp` task for codefresh only, I'm currently having a look at this as it raised a question regarding the `copy` task in #939.

About my flow:

* I use `grunt run-dev` locally and that's pretty much all I need
* I never use the `grunt release` task but I do use `grunt release platform-arch` (I sometimes use `grunt release-linux-amd64` locally to create an image that I'll host in integration environments)
* I use the `build.sh` script on release only to create the artifacts/images (then manual upload/push)
* I'm using codefresh to automatically build a latest image of the develop branch (linux amd64 only, there is no point to have multiple platform support for this at the moment)

As a later evolution, I'll probably use codefresh for the release pipeline, but that's not here yet.

1138-4EB commented:
I just rebased on top of #939. However, I'll rework this PR a little to better adapt them to your flow. So, you may focus on the other one, which is nearly done.

deviantony commented:
Something that we need to ensure will be working: https://github.com/portainer/portainer/issues/1010

deviantony commented:
#939 & #994 are now merged, let's tackle this one. Could you rebase please?

1138-4EB commented:
#1309 fixes #1286, but still portainer and docker binaries are served. All the assets should be moved to a subdir, say "app". Should some of the assets not be served, but used internally by portainer, a second level might be created: "app/public" and "app/private".

1138-4EB commented:
When bower and npm are dropped to use yarn (#1399), an offline mirror should be used, instead of saving node_modelus in the factory: https://yarnpkg.com/blog/2016/11/24/offline-mirror/
